subsys/net/l2/ethernet/arp.c

```diff
@@ -1022,23 +1014,33 @@ static enum net_verdict arp_recv(struct net_if *iface,
 				 uint16_t ptype,
 				 struct net_pkt *pkt)
 {
-	struct net_eth_hdr *hdr = NET_ETH_HDR(pkt);
+	struct net_eth_hdr hdr;
 
 	ARG_UNUSED(iface);
-	ARG_UNUSED(ptype);
+
+	/* Construct a simulated Ethernet header as we cannot access
+	 * the real Ethernet header directly any more because the caller
+	 * has pulled the Ethernet header and thus moved the buf->data
+	 * pointer.
+	 */
+	memcpy(hdr.dst.addr, net_pkt_lladdr_dst(pkt)->addr,
+	       sizeof(struct net_eth_addr));
+	memcpy(hdr.src.addr, net_pkt_lladdr_src(pkt)->addr,
+	       sizeof(struct net_eth_addr));
```

Comment:
As was suggested by @clamattia, why can't the `net_pkt_lladdr_dst/src(pkt)` functions be used instead?

Reply:
Hmm, sure that is certainly possible, I just wanted to keep changes minimal but we can change the tests too that call the arp input function.

Reply:
The changes are a bit bigger but at least we do not need to create the pseudo Ethernet header and copy data to it in this solution.